### PR TITLE
Fix shutdown

### PR DESCRIPTION
**Pull request by filmor**
### What does this implement/fix? Explain your changes.

Fixes issue #262 by countering two memory corruption issues.

### Does this close any currently open issues?

#262.

### Any other comments?

Is currently based on my other PR, but doesn't strictly need it, I could rebase.

### CODE REVIEW

**Comment by vmuriart** [comment]
Is this branch contain the same changes from #341? I'm reading it from my phone so hard to tell.

**Comment by vmuriart** [comment]
To answer my own question, it is. I'm testing this now but its looking good!

**Comment by vmuriart** [comment]
Failing on Appveyor Py33

https://ci.appveyor.com/project/vmuriart/pythonnet/build/fix-shutdown-517/job/adnsiyd4278ypy1m

**Comment by vmuriart** [comment]
@filmor you ok with me rebasing your work ontop of the master branch and push it here?

File: src/runtime/pythonengine.cs [hunk 1]
```diff
@@ -411,33 +411,54 @@ public static PyObject ModuleFromString(string name, string code)
...
+                IntPtr result = Runtime.PyRun_String(
+                    code, flag, globals.Value, locals.Value
+                    );
+
+                if (Runtime.PyErr_Occurred() != 0)
+                {
+                    throw new PythonException();
+                }
+
+                return new PyObject(result);
```
Comment (vmuriart) [inline src/runtime/pythonengine.cs]: The return within the `try` statement can get tricky. 
http://stackoverflow.com/questions/3216046/does-the-c-sharp-finally-block-always-execute

Comment (filmor) [inline src/runtime/pythonengine.cs]: I don't really see the issue here.

Comment (filmor) [inline src/runtime/pythonengine.cs]: To elaborate, the link that you posted warns against relying on `try {} finally {}` as the `finally` may not be executed, but in such a case the problem has either exited already or is in the process of doing so, so we don't have to worry about dereferencing the Python resources anyhow.

Comment (vmuriart) [inline src/runtime/pythonengine.cs]: It was more of a caveat, return statements within `try` statements confuse me at times.

File: src/runtime/importhook.cs [hunk 2]
```diff
@@ -70,25 +76,25 @@ internal static void Initialize()
...
-            }
 #endif
-            if (0 != Runtime.Py_IsInitialized())
-            {
+                Runtime.XDecref(root.pyHandle);
                 Runtime.XDecref(py_import);
             }
         }
 
+        internal static void Cleanup()
```
Comment (vmuriart) [inline src/runtime/importhook.cs]: Is there a call to `Cleanup` missing?

Comment (filmor) [inline src/runtime/importhook.cs]: That part is not used anymore as there is only a single ModuleDef instance that is kept for the whole run time, I'll remove it.

Comment (vmuriart) [inline src/runtime/importhook.cs]: Cool, after it's removed I'm ok merging it. This fixes the issue for Python2.7, but it still occassionally fails on Python3 but it seems to be due to a different section of code

Comment (filmor) [inline src/runtime/importhook.cs]: Yeah, I noticed that too, I'll try to reproduce the issue, for me it happens (after these patches) not anymore on Shutdown but only on Initialize.

**Comment by vmuriart** [comment]
Cool, will merge after CI completes. Thanks!

**Pull request merged**